File: kernel/drivers/virtio_input.c
// virtio_input.c — minimal polled virtio-input (keyboard) driver.
//
// Gives the graphical QEMU window (-device virtio-keyboard-device) a keyboard:
// the virt board exposes virtio devices over virtio-mmio (32 slots at
// 0x0A000000). We find the input device, set up its event virtqueue with a few
// receive buffers, and POLL the used ring — no IRQ wiring. virtio_kbd_getchar()
// decodes evdev key presses into ASCII; the kernel drains it on each timer tick
// and feeds the tty (kernel/main.swift), so typing reaches the busybox shell.
//
// The virtqueue rings live in cacheable RAM that QEMU's device reads/writes by
// DMA, so we clean what we write and invalidate what the device writes. (Under
// TCG these are effectively no-ops; under a caching accelerator they are not.)

#include <stdint.h>

#define VIRTIO_MMIO_BASE   0x0A000000ULL
#define VIRTIO_MMIO_STRIDE 0x200
#define VIRTIO_MMIO_COUNT  32
#define VIRTIO_ID_INPUT    18

// virtio-mmio register offsets.
#define R_MAGIC      0x000
#define R_VERSION    0x004
#define R_DEVID      0x008
#define R_DRVFEAT    0x020
#define R_DRVFEATSEL 0x024
#define R_QSEL       0x030
#define R_QNUMMAX    0x034
#define R_QNUM       0x038
#define R_QREADY     0x044
#define R_QNOTIFY    0x050
#define R_ISTATUS    0x060
#define R_IACK       0x064
#define R_STATUS     0x070
#define R_QDESCL     0x080
#define R_QDESCH     0x084
#define R_QDRVL      0x090
#define R_QDRVH      0x094
#define R_QDEVL      0x0a0
#define R_QDEVH      0x0a4

#define S_ACK    1
#define S_DRV    2
#define S_DRVOK  4
#define S_FEATOK 8

#define QSZ 8
#define VIRTQ_DESC_F_WRITE 2

struct vq_desc { uint64_t addr; uint32_t len; uint16_t flags; uint16_t next; };
struct vq_avail { uint16_t flags; uint16_t idx; uint16_t ring[QSZ]; uint16_t used_event; };
struct vq_used_elem { uint32_t id; uint32_t len; };
struct vq_used { uint16_t flags; uint16_t idx; struct vq_used_elem ring[QSZ]; uint16_t avail_event; };

static volatile uint8_t *g_mmio = 0;
static struct vq_desc  g_desc[QSZ]  __attribute__((aligned(64)));
static struct vq_avail g_avail      __attribute__((aligned(64)));
static struct vq_used  g_used       __attribute__((aligned(64)));
static uint8_t         g_evbuf[QSZ][8] __attribute__((aligned(64)));
static uint32_t g_qn = 0;
static uint16_t g_last_used = 0;

static uint8_t g_pending[64];
static int g_phead = 0, g_ptail = 0;
static int g_shift = 0;

static inline void w32(uint32_t off, uint32_t v) { *(volatile uint32_t *)(g_mmio + off) = v; }
static inline uint32_t r32(uint32_t off) { return *(volatile uint32_t *)(g_mmio + off); }

static void clean(void *p, uint64_t n) {
    uint64_t a = (uint64_t)(uintptr_t)p & ~63ULL;
    for (; a < (uint64_t)(uintptr_t)p + n; a += 64) {
        __asm__ volatile("dc cvac, %0" :: "r"(a) : "memory");
    }
    __asm__ volatile("dsb sy" ::: "memory");
}

static void invalidate(void *p, uint64_t n) {
    __asm__ volatile("dsb sy" ::: "memory");
    uint64_t a = (uint64_t)(uintptr_t)p & ~63ULL;
    for (; a < (uint64_t)(uintptr_t)p + n; a += 64) {
        __asm__ volatile("dc ivac, %0" :: "r"(a) : "memory");
    }
    __asm__ volatile("dsb sy" ::: "memory");
}

// US-keyboard evdev keycode -> ASCII (unshifted and shifted).
static const char km[128] = {
    [1]=27, [2]='1',[3]='2',[4]='3',[5]='4',[6]='5',[7]='6',[8]='7',[9]='8',[10]='9',[11]='0',
    [12]='-',[13]='=',[14]=8,[15]='\t',
    [16]='q',[17]='w',[18]='e',[19]='r',[20]='t',[21]='y',[22]='u',[23]='i',[24]='o',[25]='p',
    [26]='[',[27]=']',[28]='\r',
    [30]='a',[31]='s',[32]='d',[33]='f',[34]='g',[35]='h',[36]='j',[37]='k',[38]='l',[39]=';',[40]='\'',[41]='`',
    [43]='\\',[44]='z',[45]='x',[46]='c',[47]='v',[48]='b',[49]='n',[50]='m',[51]=',',[52]='.',[53]='/',
    [57]=' ',
};
static const char km_shift[128] = {
    [2]='!',[3]='@',[4]='#',[5]='$',[6]='%',[7]='^',[8]='&',[9]='*',[10]='(',[11]=')',
    [12]='_',[13]='+',[14]=8,[15]='\t',
    [16]='Q',[17]='W',[18]='E',[19]='R',[20]='T',[21]='Y',[22]='U',[23]='I',[24]='O',[25]='P',
    [26]='{',[27]='}',[28]='\r',
    [30]='A',[31]='S',[32]='D',[33]='F',[34]='G',[35]='H',[36]='J',[37]='K',[38]='L',[39]=':',[40]='"',[41]='~',
    [43]='|',[44]='Z',[45]='X',[46]='C',[47]='V',[48]='B',[49]='N',[50]='M',[51]='<',[52]='>',[53]='?',
    [57]=' ',
};

static uint32_t g_dbg_version = 0;

// Returns a small diagnostic code: 0 = no input device found, otherwise
// version*1000 + queue_size once the queue is up.
int virtio_kbd_init(void) {
    for (int i = 0; i < VIRTIO_MMIO_COUNT; i++) {
        volatile uint8_t *m = (volatile uint8_t *)(VIRTIO_MMIO_BASE + (uint64_t)i * VIRTIO_MMIO_STRIDE);
        if (*(volatile uint32_t *)(m + R_MAGIC) != 0x74726976) continue; // "virt"
        if (*(volatile uint32_t *)(m + R_DEVID) != VIRTIO_ID_INPUT) continue;
        g_mmio = m;
        break;
    }
    if (!g_mmio) return 0;
    g_dbg_version = r32(R_VERSION);

    w32(R_STATUS, 0);                  // reset
    w32(R_STATUS, S_ACK);
    w32(R_STATUS, S_ACK | S_DRV);
    w32(R_DRVFEATSEL, 1); w32(R_DRVFEAT, 1); // accept VIRTIO_F_VERSION_1 (bit 32)
    w32(R_DRVFEATSEL, 0); w32(R_DRVFEAT, 0);
    w32(R_STATUS, S_ACK | S_DRV | S_FEATOK);
    if (!(r32(R_STATUS) & S_FEATOK)) { g_mmio = 0; return -1; }

    w32(R_QSEL, 0);
    uint32_t max = r32(R_QNUMMAX);
    if (max == 0) { g_mmio = 0; return -2; }
    g_qn = max < QSZ ? max : QSZ;
    w32(R_QNUM, g_qn);

    for (uint32_t j = 0; j < g_qn; j++) {
        g_desc[j].addr = (uint64_t)(uintptr_t)g_evbuf[j];
        g_desc[j].len = 8;
        g_desc[j].flags = VIRTQ_DESC_F_WRITE; // device writes events into it
        g_desc[j].next = 0;
    }
    g_avail.flags = 0; g_avail.idx = 0;
    g_used.flags = 0; g_used.idx = 0;
    clean(g_desc, sizeof(g_desc));
    clean(&g_used, sizeof(g_used));

    uint64_t da = (uint64_t)(uintptr_t)g_desc;
    uint64_t aa = (uint64_t)(uintptr_t)&g_avail;
    uint64_t ua = (uint64_t)(uintptr_t)&g_used;
    w32(R_QDESCL, (uint32_t)da); w32(R_QDESCH, (uint32_t)(da >> 32));
    w32(R_QDRVL,  (uint32_t)aa); w32(R_QDRVH,  (uint32_t)(aa >> 32));
    w32(R_QDEVL,  (uint32_t)ua); w32(R_QDEVH,  (uint32_t)(ua >> 32));
    w32(R_QREADY, 1);
    w32(R_STATUS, S_ACK | S_DRV | S_FEATOK | S_DRVOK);

    // Offer every buffer to the device.
    for (uint32_t j = 0; j < g_qn; j++) g_avail.ring[j] = (uint16_t)j;
    g_avail.idx = (uint16_t)g_qn;
    clean(&g_avail, sizeof(g_avail));
    w32(R_QNOTIFY, 0);
    return (int)(g_dbg_version * 1000 + g_qn);
}

static void pump(void) {
    if (!g_mmio) return;
    invalidate(&g_used, sizeof(g_used));
    uint16_t uidx = g_used.idx;
    while (g_last_used != uidx) {
        uint32_t id = g_used.ring[g_last_used % g_qn].id % g_qn;
        invalidate(g_evbuf[id], 8);
        uint16_t type = (uint16_t)(g_evbuf[id][0] | (g_evbuf[id][1] << 8));
        uint16_t code = (uint16_t)(g_evbuf[id][2] | (g_evbuf[id][3] << 8));
        uint32_t value = (uint32_t)(g_evbuf[id][4] | (g_evbuf[id][5] << 8) |
                                    (g_evbuf[id][6] << 16) | (g_evbuf[id][7] << 24));
        if (type == 1) { // EV_KEY
            if (code == 42 || code == 54) {        // L/R shift
                g_shift = (value != 0);
            } else if (value == 1 || value == 2) { // press or auto-repeat
                char c = g_shift ? km_shift[code & 0x7f] : km[code & 0x7f];
                if (c) {
                    int nt = (g_ptail + 1) % 64;
                    if (nt != g_phead) { g_pending[g_ptail] = (uint8_t)c; g_ptail = nt; }
                }
            }
        }
        // Hand the buffer back to the device.
        g_avail.ring[g_avail.idx % g_qn] = (uint16_t)id;
        g_avail.idx++;
        clean(&g_avail, sizeof(g_avail));
        g_last_used++;
    }
    w32(R_QNOTIFY, 0);
    uint32_t is = r32(R_ISTATUS);
    if (is) w32(R_IACK, is); // keep the device quiet even though we poll
}

// Return the next typed ASCII byte, or -1 if nothing is queued.
int virtio_kbd_getchar(void) {
    pump();
    if (g_phead == g_ptail) return -1;
    uint8_t b = g_pending[g_phead];
    g_phead = (g_phead + 1) % 64;
    return (int)b;
}

File: kernel/drivers/virtio_input.swift
// SPDX-License-Identifier: Apache-2.0
// virtio_input.swift — minimal polled virtio-input (keyboard) driver.
//
// Gives the graphical QEMU window (-device virtio-keyboard-device) a keyboard:
// the virt board exposes virtio devices over virtio-mmio (32 slots at
// 0x0A000000). We find the input device, set up its event virtqueue with a few
// receive buffers, and POLL the used ring — no IRQ wiring. virtioKbdGetchar()
// decodes evdev key presses into ASCII; the kernel drains it on each timer tick
// and feeds the tty (kernel/main.swift), so typing reaches the busybox shell.
// Arrow and navigation keys (Left/Right/Up/Down/Home/End/Delete) are emitted as
// the ANSI escape sequences the tty's cooked line editor decodes for cursor
// movement and mid-line editing.
//
// Swift rewrite of the former virtio_input.c, following virtio_net.swift: the
// ring and event buffers are one PMM page, with the small ASCII pending queue
// living (CPU-only, never DMA'd) in the page's spare area. MMIO and cache
// maintenance go through the io.h C bridge; under TCG the clean/invalidate are
// effectively no-ops, under a caching accelerator they are not.

private let KBD_MMIO_BASE: UInt   = 0x0A00_0000
private let KBD_MMIO_STRIDE: UInt = 0x200
private let KBD_MMIO_COUNT = 32
private let VIRTIO_ID_INPUT: UInt32 = 18

// virtio-mmio register offsets.
private let R_MAGIC: UInt      = 0x000
private let R_VERSION: UInt    = 0x004
private let R_DEVID: UInt      = 0x008
private let R_DRVFEAT: UInt    = 0x020
private let R_DRVFEATSEL: UInt = 0x024
private let R_QSEL: UInt       = 0x030
private let R_QNUMMAX: UInt    = 0x034
private let R_QNUM: UInt       = 0x038
private let R_QREADY: UInt     = 0x044
private let R_QNOTIFY: UInt    = 0x050
private let R_ISTATUS: UInt    = 0x060
private let R_IACK: UInt       = 0x064
private let R_STATUS: UInt     = 0x070
private let R_QDESCL: UInt     = 0x080
private let R_QDESCH: UInt     = 0x084
private let R_QDRVL: UInt      = 0x090
private let R_QDRVH: UInt      = 0x094
private let R_QDEVL: UInt      = 0x0a0
private let R_QDEVH: UInt      = 0x0a4

private let VIRTIO_MAGIC: UInt32 = 0x74726976   // "virt"

private let S_ACK: UInt32    = 1
private let S_DRV: UInt32    = 2
private let S_DRVOK: UInt32  = 4
private let S_FEATOK: UInt32 = 8

private let KBD_QSZ = 8
private let VIRTQ_DESC_F_WRITE: UInt16 = 2

// Single-page layout: rings as in virtio_net.swift, plus the 8x8 event-buffer
// pool and a 64-byte CPU-only pending ring in the page's spare area.
private let OFF_DESC: UInt    = 0x000
private let OFF_AVAIL: UInt   = 0x080
private let OFF_USED: UInt    = 0x100
private let OFF_EVBUF: UInt   = 0x200  // KBD_QSZ x 8 bytes, device-write (DMA)
private let OFF_PENDING: UInt = 0x300  // 64-byte ASCII ring, never DMA'd

private var kbdMmio: UInt = 0
private var kbdRingBase: UInt = 0
private var kbdQn: UInt32 = 0
private var kbdAvailIdx: UInt16 = 0
private var kbdLastUsed: UInt16 = 0
private var kbdShift = false
private var kbdPHead = 0
private var kbdPTail = 0
private var kbdDbgVersion: UInt32 = 0

// US-keyboard evdev keycode -> ASCII, filled once at init (mirrors the km /
// km_shift tables in the former virtio_input.c). Index by keycode & 0x7f.
private var kbdKm = [UInt8](repeating: 0, count: 128)
private var kbdKmShift = [UInt8](repeating: 0, count: 128)
private var kbdKmReady = false

private func kbdFillKeymaps() {
    if kbdKmReady { return }
    func put(_ start: Int, _ s: StaticString, shift: Bool) {
        s.withUTF8Buffer { b in
            var k = 0
            while k < b.count {
                if shift { kbdKmShift[start + k] = b[k] } else { kbdKm[start + k] = b[k] }
                k += 1
            }
        }
    }
    // Unshifted.
    put(2, "1234567890-=", shift: false)
    kbdKm[14] = 8; kbdKm[15] = 9                       // backspace, tab
    put(16, "qwertyuiop[]", shift: false)
    kbdKm[28] = 13                                     // enter
    put(30, "asdfghjkl;'`", shift: false)
    kbdKm[43] = UInt8(ascii: "\\")
    put(44, "zxcvbnm,./", shift: false)
    kbdKm[57] = 32                                     // space
    // Shifted.
    put(2, "!@#$%^&*()_+", shift: true)
    kbdKmShift[14] = 8; kbdKmShift[15] = 9
    put(16, "QWERTYUIOP{}", shift: true)
    kbdKmShift[28] = 13
    put(30, "ASDFGHJKL:\"~", shift: true)
    kbdKmShift[43] = UInt8(ascii: "|")
    put(44, "ZXCVBNM<>?", shift: true)
    kbdKmShift[57] = 32
    kbdKmReady = true
}

// --- cache maintenance ------------------------------------------------------
private func kbdClean(_ pa: UInt, _ n: Int) {
    var a = pa & ~UInt(63)
    let end = pa + UInt(n)
    while a < end { dc_cvac(a); a += 64 }
    dsb_sy()
}
private func kbdInvalidate(_ pa: UInt, _ n: Int) {
    dsb_sy()
    var a = pa & ~UInt(63)
    let end = pa + UInt(n)
    while a < end { dc_ivac(a); a += 64 }
    dsb_sy()
}
private func kbdZeroPage(_ pa: UInt) {
    let p = UnsafeMutableRawPointer(bitPattern: pa)!
    var i = 0
    while i < 4096 { p.storeBytes(of: UInt8(0), toByteOffset: i, as: UInt8.self); i += 1 }
}

// --- virtqueue accessors ----------------------------------------------------
private func kbdDescSet(_ i: Int, addr: UInt64, len: UInt32, flags: UInt16, next: UInt16) {
    let d = UnsafeMutableRawPointer(bitPattern: kbdRingBase + OFF_DESC + UInt(i * 16))!
    d.storeBytes(of: addr, toByteOffset: 0, as: UInt64.self)
    d.storeBytes(of: len, toByteOffset: 8, as: UInt32.self)
    d.storeBytes(of: flags, toByteOffset: 12, as: UInt16.self)
    d.storeBytes(of: next, toByteOffset: 14, as: UInt16.self)
}
private func kbdAvailAdd(descIdx: UInt16) {
    let avail = UnsafeMutableRawPointer(bitPattern: kbdRingBase + OFF_AVAIL)!
    let slot = Int(kbdAvailIdx % UInt16(kbdQn))
    avail.storeBytes(of: descIdx, toByteOffset: 4 + slot * 2, as: UInt16.self)
    kbdAvailIdx &+= 1
    avail.storeBytes(of: kbdAvailIdx, toByteOffset: 2, as: UInt16.self)
}
private func kbdUsedIdx() -> UInt16 {
    UnsafeRawPointer(bitPattern: kbdRingBase + OFF_USED)!.load(fromByteOffset: 2, as: UInt16.self)
}
private func kbdUsedElemId(_ slot: Int) -> UInt32 {
    UnsafeRawPointer(bitPattern: kbdRingBase + OFF_USED + 4 + UInt(slot * 8))!.load(fromByteOffset: 0, as: UInt32.self)
}

// --- pending ASCII ring (CPU-only, in the page's spare area) ----------------
private func kbdPushByte(_ c: UInt8) {
    let nt = (kbdPTail + 1) % 64
    if nt != kbdPHead {
        UnsafeMutableRawPointer(bitPattern: kbdRingBase + OFF_PENDING)!
            .storeBytes(of: c, toByteOffset: kbdPTail, as: UInt8.self)
        kbdPTail = nt
    }
}
private func kbdPushSeq(_ s: StaticString) {
    s.withUTF8Buffer { b in for x in b { kbdPushByte(x) } }
}

// Returns a small diagnostic code: 0 = no input device found, otherwise
// version*1000 + queue_size once the queue is up.
func virtioKbdInit() -> Int32 {
    kbdMmio = 0
    var found: UInt = 0
    var i = 0
    while i < KBD_MMIO_COUNT {
        let m = KBD_MMIO_BASE + UInt(i) * KBD_MMIO_STRIDE
        i += 1
        if mmio_read32(m + R_MAGIC) != VIRTIO_MAGIC { continue }
        if mmio_read32(m + R_DEVID) != VIRTIO_ID_INPUT { continue }
        found = m
        break
    }
    if found == 0 { return 0 }
    kbdMmio = found
    kbdDbgVersion = mmio_read32(found + R_VERSION)

    if kbdRingBase == 0 {
        let r = pmm_alloc_page()
        if r == 0 { kbdMmio = 0; return -3 }
        kbdRingBase = r
    }
    kbdZeroPage(kbdRingBase)
    kbdFillKeymaps()
    kbdAvailIdx = 0
    kbdLastUsed = 0
    kbdPHead = 0
    kbdPTail = 0
    kbdShift = false

    mmio_write32(found + R_STATUS, 0)                  // reset
    mmio_write32(found + R_STATUS, S_ACK)
    mmio_write32(found + R_STATUS, S_ACK | S_DRV)
    mmio_write32(found + R_DRVFEATSEL, 1); mmio_write32(found + R_DRVFEAT, 1) // VIRTIO_F_VERSION_1
    mmio_write32(found + R_DRVFEATSEL, 0); mmio_write32(found + R_DRVFEAT, 0)
    mmio_write32(found + R_STATUS, S_ACK | S_DRV | S_FEATOK)
    if (mmio_read32(found + R_STATUS) & S_FEATOK) == 0 { kbdMmio = 0; return -1 }

    mmio_write32(found + R_QSEL, 0)
    let maxq = mmio_read32(found + R_QNUMMAX)
    if maxq == 0 { kbdMmio = 0; return -2 }
    kbdQn = maxq < UInt32(KBD_QSZ) ? maxq : UInt32(KBD_QSZ)
    mmio_write32(found + R_QNUM, kbdQn)

    var j: UInt32 = 0
    while j < kbdQn {
        let buf = kbdRingBase + OFF_EVBUF + UInt(Int(j) * 8)
        kbdDescSet(Int(j), addr: UInt64(buf), len: 8, flags: VIRTQ_DESC_F_WRITE, next: 0) // device writes events
        j += 1
    }
    kbdClean(kbdRingBase + OFF_DESC, KBD_QSZ * 16)
    kbdClean(kbdRingBase + OFF_USED, 72)

    let da = UInt64(kbdRingBase + OFF_DESC)
    let aa = UInt64(kbdRingBase + OFF_AVAIL)
    let ua = UInt64(kbdRingBase + OFF_USED)
    mmio_write32(found + R_QDESCL, UInt32(da & 0xFFFF_FFFF)); mmio_write32(found + R_QDESCH, UInt32(da >> 32))
    mmio_write32(found + R_QDRVL,  UInt32(aa & 0xFFFF_FFFF)); mmio_write32(found + R_QDRVH,  UInt32(aa >> 32))
    mmio_write32(found + R_QDEVL,  UInt32(ua & 0xFFFF_FFFF)); mmio_write32(found + R_QDEVH,  UInt32(ua >> 32))
    mmio_write32(found + R_QREADY, 1)
    mmio_write32(found + R_STATUS, S_ACK | S_DRV | S_FEATOK | S_DRVOK)

    // Offer every buffer to the device.
    let avail = UnsafeMutableRawPointer(bitPattern: kbdRingBase + OFF_AVAIL)!
    j = 0
    while j < kbdQn { avail.storeBytes(of: UInt16(j), toByteOffset: 4 + Int(j) * 2, as: UInt16.self); j += 1 }
    kbdAvailIdx = UInt16(kbdQn)
    avail.storeBytes(of: kbdAvailIdx, toByteOffset: 2, as: UInt16.self)
    kbdClean(kbdRingBase + OFF_AVAIL, 32)
    mmio_write32(found + R_QNOTIFY, 0)
    return Int32(kbdDbgVersion * 1000 + kbdQn)
}

private func kbdPump() {
    if kbdMmio == 0 { return }
    kbdInvalidate(kbdRingBase + OFF_USED, 72)
    let uidx = kbdUsedIdx()
    while kbdLastUsed != uidx {
        let id = kbdUsedElemId(Int(kbdLastUsed % UInt16(kbdQn))) % kbdQn
        let evb = kbdRingBase + OFF_EVBUF + UInt(Int(id) * 8)
        kbdInvalidate(evb, 8)
        let p = UnsafeRawPointer(bitPattern: evb)!
        let type  = UInt16(p.load(fromByteOffset: 0, as: UInt8.self)) | (UInt16(p.load(fromByteOffset: 1, as: UInt8.self)) << 8)
        let code  = UInt16(p.load(fromByteOffset: 2, as: UInt8.self)) | (UInt16(p.load(fromByteOffset: 3, as: UInt8.self)) << 8)
        let value = UInt32(p.load(fromByteOffset: 4, as: UInt8.self)) | (UInt32(p.load(fromByteOffset: 5, as: UInt8.self)) << 8)
                  | (UInt32(p.load(fromByteOffset: 6, as: UInt8.self)) << 16) | (UInt32(p.load(fromByteOffset: 7, as: UInt8.self)) << 24)
        if type == 1 { // EV_KEY
            if code == 42 || code == 54 {            // L/R shift
                kbdShift = (value != 0)
            } else if value == 1 || value == 2 {     // press or auto-repeat
                // Navigation keys become the ANSI escape sequences the tty line
                // editor decodes; everything else is a single ASCII byte.
                switch code {
                case 103: kbdPushSeq("\u{1b}[A")     // Up
                case 108: kbdPushSeq("\u{1b}[B")     // Down
                case 106: kbdPushSeq("\u{1b}[C")     // Right
                case 105: kbdPushSeq("\u{1b}[D")     // Left
                case 102: kbdPushSeq("\u{1b}[H")     // Home
                case 107: kbdPushSeq("\u{1b}[F")     // End
                case 111: kbdPushSeq("\u{1b}[3~")    // Delete (forward)
                default:
                    let c = kbdShift ? kbdKmShift[Int(code & 0x7f)] : kbdKm[Int(code & 0x7f)]
                    if c != 0 { kbdPushByte(c) }
                }
            }
        }
        // Hand the buffer back to the device.
        kbdAvailAdd(descIdx: UInt16(id))
        kbdClean(kbdRingBase + OFF_AVAIL, 32)
        kbdLastUsed &+= 1
    }
    mmio_write32(kbdMmio + R_QNOTIFY, 0)
    let ist = mmio_read32(kbdMmio + R_ISTATUS)
    if ist != 0 { mmio_write32(kbdMmio + R_IACK, ist) } // keep the device quiet even though we poll
}

// Return the next typed ASCII byte, or -1 if nothing is queued.
func virtioKbdGetchar() -> Int32 {
    kbdPump()
    if kbdPHead == kbdPTail { return -1 }
    let pending = UnsafeRawPointer(bitPattern: kbdRingBase + OFF_PENDING)!
    let b = pending.load(fromByteOffset: kbdPHead, as: UInt8.self)
    kbdPHead = (kbdPHead + 1) % 64
    return Int32(b)
}
